File: src/training.ts
import type { PhotoType } from "./analysis";
import type { AssetBatchContext } from "./batch-context";
import type { EditRecipe } from "./recipe";

export type PairValidationStatus =
  "pending" | "ready" | "needs_review" | "rejected" | "unusable";
export type GeometryRelationship =
  | "unknown"
  | "exact_or_near"
  | "crop_difference"
  | "major_mismatch"
  | "unusable";
export type TrainingSplit = "unassigned" | "train" | "validation" | "excluded";
export type TargetFitConfidence = "high" | "medium" | "low";
export type ValidationFeedback = "accept" | "needs_adjustment" | "reject";

export interface PairValidation {
  status: PairValidationStatus;
  geometry: GeometryRelationship;
  structural_similarity: number | null;
  source_width: number | null;
  source_height: number | null;
  reference_width: number | null;
  reference_height: number | null;
  diagnostics: string[];
}

export interface CreativeControls {
  exposure_ev: number;
  temperature_delta: number;
  tint: number;
  contrast: number;
  highlights: number;
  shadows: number;
  whites: number;
  blacks: number;
  saturation: number;
  vibrance: number;
  texture: number;
  clarity: number;
  dehaze: number;
  sharpening_amount: number;
  noise_reduction: number;
  vignette_amount: number;
}

export interface TargetRecipeResult {
  schema_version: number;
  optimizer_version: string;
  cache_identity: string;
  recipe: EditRecipe;
  controls: CreativeControls;
  confidence: TargetFitConfidence;
  loss: {
    total: number;
    luminance: number;
    color_balance: number;
    saturation: number;
    structure: number;
  };
  iterations: number;
  unsupported_differences: string[];
  warnings: string[];
}

export interface TrainingPair {
  schema_version: number;
  pair_id: string;
  dataset_id: string;
  source_job_id: string;
  source_asset_id: string;
  source_path: string;
  reference_path: string;
  photo_type: PhotoType;
  source_fingerprint: string;
  reference_fingerprint: string;
  validation: PairValidation;
  source_analysis_id: string | null;
  batch_context: AssetBatchContext | null;
  scene_group_id: string | null;
  target: TargetRecipeResult | null;
  split: TrainingSplit;
  excluded: boolean;
  feedback: ValidationFeedback | null;
  diagnostics: string[];
}

export interface TrainingDataset {
  schema_version: number;
  dataset_id: string;
  job_id: string;
  style_name: string;
  photo_type: PhotoType;
  pairs: TrainingPair[];
  created_at: string;
  updated_at: string;
  dataset_fingerprint: string | null;
  feature_schema: string;
  renderer_version: string;
  target_recipe_schema: number;
  batch_context_id: string | null;
  warnings: string[];
  before_files: string[];
  after_files: string[];
  alignment: TrainingAlignment | null;
}

export interface TrainingAlignment {
  before_count: number;
  after_count: number;
  matched_count: number;
  ambiguous_count: number;
  unmatched_before: string[];
  unmatched_after: string[];
  first_before: string | null;
  first_after: string | null;
  last_before: string | null;
  last_after: string | null;
  start_aligned: boolean;
  end_aligned: boolean;
  order_fallback_used: boolean;
  diagnostics: string[];
}

export interface TrainingConfig {
  validation_percent: number;
  regularization: number;
  learning_rate: number;
  epochs: number;
  exclude_low_confidence: boolean;
  deterministic: boolean;
}

export const DEFAULT_TRAINING_CONFIG: TrainingConfig = {
  validation_percent: 20,
  regularization: 0.08,
  learning_rate: 0.025,
  epochs: 1200,
  exclude_low_confidence: true,
  deterministic: true,
};

export type TrainingRunStatus =
  "queued" | "running" | "complete" | "failed" | "cancelled" | "interrupted";
export type TrainingStage =
  | "queued"
  | "validating_pairs"
  | "analyzing"
  | "estimating_target_recipes"
  | "building_examples"
  | "training"
  | "validating"
  | "exporting_style"
  | "complete"
  | "stopped";

export interface MetricSet {
  recipe_mae: Record<string, number>;
  mean_recipe_mae: number;
  rendered_loss: number | null;
}

export interface TrainingMetrics {
  train: MetricSet;
  validation: MetricSet;
  neutral_baseline: MetricSet;
  mean_baseline: MetricSet;
  beats_mean_baseline: boolean;
  overfitting_warning: string | null;
  warnings: string[];
}

export interface TrainingRun {
  schema_version: number;
  run_id: string;
  dataset_id: string;
  style_id: string | null;
  style_name: string;
  style_version: string | null;
  status: TrainingRunStatus;
  stage: TrainingStage;
  completed: number;
  total: number;
  config: TrainingConfig;
  metrics: TrainingMetrics | null;
  artifact_path: string | null;
  started_at: string;
  updated_at: string;
  duration_ms: number;
  error: string | null;
}

export interface AutoMatchResult {
  dataset: TrainingDataset;
  matching: {
    matched: Array<{
      source_asset_id: string;
      source_filename: string;
      source_path?: string;
      reference_path: string;
    }>;
    ambiguous_sources: string[];
    unmatched_references: string[];
    unmatched_sources?: string[];
    before_count?: number;
    after_count?: number;
    start_aligned?: boolean;
    end_aligned?: boolean;
    order_fallback_used?: boolean;
    diagnostics?: string[];
  };
}

export interface TrainingPreviewSet {
  source_data: string;
  ai_data: string | null;
  target_data: string | null;
  reference_data: string;
}

export interface MatchingProgress {
  request_id: string;
  dataset_id: string;
  status: "running" | "complete" | "failed" | "cancelled";
  stage: string;
  processed: number;
  total: number;
  error: string | null;
}

export interface ValidationEditingSelection {
  photo_type: PhotoType;
  asset_ids: string[];
}

File: crates/photo-contracts/src/training.rs
//! Local Training Studio contracts. Training learns creative recipe controls, never pixels.
use crate::{
    analysis::PhotoType,
    batch_context::AssetBatchContext,
    trained_style::{PredictedCreativeAdjustments, StyleControl, STYLE_FEATURE_SCHEMA_V1},
    EditRecipe, RECIPE_SCHEMA_VERSION,
};
use serde::{Deserialize, Serialize};
use std::{collections::BTreeMap, path::PathBuf};

pub const TRAINING_PAIR_SCHEMA_VERSION: u32 = 1;
pub const TRAINING_DATASET_SCHEMA_VERSION: u32 = 1;
pub const TARGET_RECIPE_SCHEMA_VERSION: u32 = 1;
pub const TRAINING_RUN_SCHEMA_VERSION: u32 = 1;
pub const TRAINER_VERSION: &str = "regularized-linear-recipe-v1";
pub const TARGET_OPTIMIZER_VERSION: &str = "staged-photographic-proxy-v1";

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PairValidationStatus {
    #[default]
    Pending,
    Ready,
    NeedsReview,
    Rejected,
    Unusable,
}

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum GeometryRelationship {
    #[default]
    Unknown,
    ExactOrNear,
    CropDifference,
    MajorMismatch,
    Unusable,
}

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TrainingSplit {
    #[default]
    Unassigned,
    Train,
    Validation,
    Excluded,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TargetFitConfidence {
    High,
    Medium,
    Low,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ValidationFeedback {
    Accept,
    NeedsAdjustment,
    Reject,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
#[serde(default, deny_unknown_fields)]
pub struct PairValidation {
    pub status: PairValidationStatus,
    pub geometry: GeometryRelationship,
    pub structural_similarity: Option<f32>,
    pub source_width: Option<u32>,
    pub source_height: Option<u32>,
    pub reference_width: Option<u32>,
    pub reference_height: Option<u32>,
    pub diagnostics: Vec<String>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default, deny_unknown_fields)]
pub struct TrainingAlignment {
    pub before_count: u32,
    pub after_count: u32,
    pub matched_count: u32,
    pub ambiguous_count: u32,
    pub unmatched_before: Vec<PathBuf>,
    pub unmatched_after: Vec<PathBuf>,
    pub first_before: Option<PathBuf>,
    pub first_after: Option<PathBuf>,
    pub last_before: Option<PathBuf>,
    pub last_after: Option<PathBuf>,
    pub start_aligned: bool,
    pub end_aligned: bool,
    pub order_fallback_used: bool,
    pub diagnostics: Vec<String>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct TargetLossBreakdown {
    pub total: f32,
    pub luminance: f32,
    pub color_balance: f32,
    pub saturation: f32,
    pub structure: f32,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct TargetRecipeResult {
    pub schema_version: u32,
    pub optimizer_version: String,
    pub cache_identity: String,
    pub recipe: EditRecipe,
    pub controls: PredictedCreativeAdjustments,
    pub confidence: TargetFitConfidence,
    pub loss: TargetLossBreakdown,
    pub iterations: u32,
    pub unsupported_differences: Vec<String>,
    pub warnings: Vec<String>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct TrainingPair {
    pub schema_version: u32,
    pub pair_id: String,
    pub dataset_id: String,
    pub source_job_id: String,
    pub source_asset_id: String,
    pub source_path: PathBuf,
    pub reference_path: PathBuf,
    pub photo_type: PhotoType,
    pub source_fingerprint: String,
    pub reference_fingerprint: String,
    pub validation: PairValidation,
    pub source_analysis_id: Option<String>,
    pub batch_context: Option<AssetBatchContext>,
    pub scene_group_id: Option<String>,
    pub target: Option<TargetRecipeResult>,
    pub split: TrainingSplit,
    pub excluded: bool,
    pub feedback: Option<ValidationFeedback>,
    pub diagnostics: Vec<String>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct TrainingDataset {
    pub schema_version: u32,
    pub dataset_id: String,
    pub job_id: String,
    pub style_name: String,
    pub photo_type: PhotoType,
    pub pairs: Vec<TrainingPair>,
    pub created_at: String,
    pub updated_at: String,
    pub dataset_fingerprint: Option<String>,
    pub feature_schema: String,
    pub renderer_version: String,
    pub target_recipe_schema: u32,
    pub batch_context_id: Option<String>,
    pub warnings: Vec<String>,
    #[serde(default)]
    pub before_files: Vec<PathBuf>,
    #[serde(default)]
    pub after_files: Vec<PathBuf>,
    #[serde(default)]
    pub alignment: Option<TrainingAlignment>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(default, deny_unknown_fields)]
pub struct TrainingConfig {
    pub validation_percent: u8,
    pub regularization: f32,
    pub learning_rate: f32,
    pub epochs: u32,
    pub exclude_low_confidence: bool,
    pub deterministic: bool,
}

impl Default for TrainingConfig {
    fn default() -> Self {
        Self {
            validation_percent: 20,
            regularization: 0.08,
            learning_rate: 0.025,
            epochs: 1_200,
            exclude_low_confidence: true,
            deterministic: true,
        }
    }
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
#[serde(default, deny_unknown_fields)]
pub struct MetricSet {
    pub recipe_mae: BTreeMap<StyleControl, f32>,
    pub mean_recipe_mae: f32,
    pub rendered_loss: Option<f32>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct TrainingMetrics {
    pub train: MetricSet,
    pub validation: MetricSet,
    pub neutral_baseline: MetricSet,
    pub mean_baseline: MetricSet,
    pub beats_mean_baseline: bool,
    pub overfitting_warning: Option<String>,
    pub warnings: Vec<String>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TrainingRunStatus {
    Queued,
    Running,
    Complete,
    Failed,
    Cancelled,
    Interrupted,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TrainingStage {
    Queued,
    ValidatingPairs,
    Analyzing,
    EstimatingTargetRecipes,
    BuildingExamples,
    Training,
    Validating,
    ExportingStyle,
    Complete,
    Stopped,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct TrainingRun {
    pub schema_version: u32,
    pub run_id: String,
    pub dataset_id: String,
    pub style_id: Option<String>,
    pub style_name: String,
    pub style_version: Option<String>,
    pub status: TrainingRunStatus,
    pub stage: TrainingStage,
    pub completed: u32,
    pub total: u32,
    pub config: TrainingConfig,
    pub metrics: Option<TrainingMetrics>,
    pub artifact_path: Option<PathBuf>,
    pub started_at: String,
    pub updated_at: String,
    pub duration_ms: u64,
    pub error: Option<String>,
}

fn digest(value: &str) -> bool {
    value.len() == 64 && value.bytes().all(|byte| byte.is_ascii_hexdigit())
}

impl TrainingPair {
    pub fn validate_shape(&self) -> Result<(), String> {
        if self.schema_version != TRAINING_PAIR_SCHEMA_VERSION {
            return Err("Unsupported training-pair schema".into());
        }
        if self.pair_id.is_empty()
            || self.dataset_id.is_empty()
            || self.source_job_id.is_empty()
            || self.source_asset_id.is_empty()
            || self.source_path.as_os_str().is_empty()
            || self.reference_path.as_os_str().is_empty()
            || (!self.source_fingerprint.is_empty() && !digest(&self.source_fingerprint))
            || (!self.reference_fingerprint.is_empty() && !digest(&self.reference_fingerprint))
        {
            return Err("Training-pair identity is invalid".into());
        }
        if let Some(target) = &self.target {
            if target.schema_version != TARGET_RECIPE_SCHEMA_VERSION
                || target.optimizer_version != TARGET_OPTIMIZER_VERSION
                || !digest(&target.cache_identity)
                || !target.loss.total.is_finite()
            {
                return Err("Target recipe result is invalid".into());
            }
            target
                .recipe
                .clone()
                .validated()
                .map_err(|error| error.message)?;
        }
        Ok(())
    }
}

impl TrainingDataset {
    pub fn validate_shape(&self) -> Result<(), String> {
        if self.schema_version != TRAINING_DATASET_SCHEMA_VERSION
            || self.dataset_id.is_empty()
            || self.job_id.is_empty()
            || self.style_name.trim().is_empty()
            || self.style_name.len() > 256
            || self.feature_schema != STYLE_FEATURE_SCHEMA_V1
            || self.renderer_version.is_empty()
            || self.target_recipe_schema != RECIPE_SCHEMA_VERSION
            || self
                .dataset_fingerprint
                .as_ref()
                .is_some_and(|value| !digest(value))
        {
            return Err("Training dataset identity or compatibility is invalid".into());
        }
        for pair in &self.pairs {
            pair.validate_shape()?;
            if pair.dataset_id != self.dataset_id
                || pair.source_job_id != self.job_id
                || pair.photo_type != self.photo_type
            {
                return Err("Training pair does not belong to this dataset".into());
            }
        }
        Ok(())
    }
}

impl TrainingConfig {
    pub fn validate(&self) -> Result<(), String> {
        if !(1..=50).contains(&self.validation_percent)
            || !self.regularization.is_finite()
            || !(0.0..=10.0).contains(&self.regularization)
            || !self.learning_rate.is_finite()
            || !(0.0001..=0.25).contains(&self.learning_rate)
            || !(10..=20_000).contains(&self.epochs)
        {
            return Err("Training configuration is outside safe bounds".into());
        }
        Ok(())
    }
}
